test/e2e/e2e_test.go

```diff
@@ -461,7 +489,36 @@ func runYtt(t *testing.T, files testInputFiles, stdinFileName string, flags yttF
 		command.Stdin = fileToUseInStdIn
 	}
 	output, err := command.Output()
+
 	require.NoError(t, err, stdError.String())
 
 	return string(output)
 }
+func runYttExpectingError(t *testing.T, files testInputFiles, stdinFileName string, flags yttFlags, envs []string) string {
+	var fileFlags []string
+	for _, file := range files {
+		fileFlags = append(fileFlags, "-f", file)
+	}
+
+	var yttFlags []string
+	for _, flagElement := range flags {
+		for flagName, flagVal := range flagElement {
+			if flagVal != "" {
+				yttFlags = append(yttFlags, flagName, flagVal)
+			} else {
+				yttFlags = append(yttFlags, flagName)
+			}
+		}
+	}
+
+	command := exec.Command("../../ytt", append(fileFlags, yttFlags...)...)
+	stdError := bytes.NewBufferString("")
+	command.Stderr = stdError
+	command.Env = append(command.Env, envs...)
+
+	_, err := command.Output()
+	require.Error(t, err, stdError.String())
+
+	return stdError.String()
+
+}
```

Comment:
Duplicating a function that _almost_ but _doesn't quite_ fit is a terrific first move in situations like this.

...

Also, given that test code isn't _itself_ covered, we tend to get better results by leaving that test subsystem a bit more "DAMP" (i.e. less Do not Repeat Yourself). 👍🏻 

So, the question arises at this point: should the common code between `runYtt()` and `runYttExpectingError()` be extracted?
In my mind, we're trying to balance "correctness" with "maintainability."
- Correctness: we can be more confident that our test code is correct when there are fewer levels of indirection — it's more straight-forward to read.
- Maintainability: the more modular our code (including minimizing duplication), the less effort it takes to affect changes.

In this very specific case, I think the core question I ask myself is:

_How fragmented are the differences between the two functions (i.e. are there differences throughout? or are they all in one place?)_

Reply:
Thanks John for the detailed review. I have refactored the code to minimize the duplication.